src/legacy/core_plugins/interpreter/types/types.ts

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * A generic type which represents a custom Expression Type definition.
+ *
+ * TODO: Add typings for the `availableTypes` parameter that's used by the from/to methods.
+ */
+export interface ExpressionType<Name extends string, Type, SerializedType = undefined> {
+  name: Name;
+  validate?: (type: any) => void | Error;
+  serialize?: (type: Type) => SerializedType;
+  deserialize?: (type: SerializedType) => Type;
+  from?: Record<string, (ctx: any, availableTypes: Record<string, any>) => Type>;
+  to?: Record<string, (type: Type, availableTypes: Record<string, any>) => unknown>;
```

Comment:
I looked at where we are actually calling from/to while casting in the interpreter, and I don’t think I have these signatures quite right — would appreciate any input (maybe @w33ble knows?)

Reply:
I'll admit I can't make a ton of sense of this typing, not knowing TS well, but I don't think you want `to` to return `unknown`... you actually *always* know the type you want back, it's implied by the object's key and if the types (our types) are all typed (typescript), then you should be able to specify the output by passing in the type (typescript).

```
to: {
  pointseries: PointseriesExpressionType => ...
}
```

I have no idea how to pass`PointseriesExpressionType` in as a generic or whatever in the `ExpressionType` definition though, or if it even makes sense to do so. Maybe that's the point of `unknown` (I don't know how `unknown` works...)?

Reply:
@w33ble notes from our offline conversation:
- Add todos around typing values for availableTypes parameters. We can likely add stronger typings for these once more parts of the interpreter have been converted to TS.
- Consider renaming to `ExpressionTypeDefinition` to avoid confusion, since these are referred to as "type definitions" elsewhere in Kibana